flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java

```diff
@@ -0,0 +1,248 @@
+		kafkaProperties.setProperty(PRODUCER_PARALLELISM, String.valueOf(producerParallelism));
+		kafkaProperties.setProperty(PARTITION_NUMBER, String.valueOf(numberOfPartitions));
+
+		StreamExecutionEnvironment env = inputStream.getExecutionEnvironment();
+		TypeInformationSerializationSchema<T> schema =
+			new TypeInformationSerializationSchema<>(inputStream.getType(), env.getConfig());
+
+		writeKeyBy(inputStream, topic, kafkaProperties, keySelector);
+		return readKeyBy(topic, env, schema, kafkaProperties, keySelector);
+	}
+
+	/**
+	 * Writes to a kafka shuffle with the partition decided by keys.
+	 *
+	 * @param inputStream Input stream to the kafka
+	 * @param topic Kafka topic
+	 * @param kafkaProperties Kafka properties
+	 * @param fields Key positions from inputStream
+	 * @param <T> Input type
+	 */
+	public static <T> void writeKeyBy(
+			DataStream<T> inputStream,
+			String topic,
+			Properties kafkaProperties,
+			int... fields) {
+		writeKeyBy(inputStream, topic, kafkaProperties, keySelector(inputStream, fields));
+	}
+
+	/**
+	 * Writes to a kafka shuffle with the partition decided by keys.
+	 *
+	 * @param inputStream Input stream to the kafka
+	 * @param topic Kafka topic
+	 * @param kafkaProperties Kafka properties
+	 * @param keySelector Key(K) based on input(T)
+	 * @param <T> Input type
+	 * @param <K> Key type
+	 */
+	public static <T, K> void writeKeyBy(
+			DataStream<T> inputStream,
+			String topic,
+			Properties kafkaProperties,
+			KeySelector<T, K> keySelector) {
+		StreamExecutionEnvironment env = inputStream.getExecutionEnvironment();
+		TypeInformationSerializationSchema<T> schema =
+			new TypeInformationSerializationSchema<>(inputStream.getType(), env.getConfig());
+
+		// write data to Kafka
+		FlinkKafkaShuffleProducer<T, K> kafkaProducer = new FlinkKafkaShuffleProducer<>(
+			topic,
+			schema.getSerializer(),
+			kafkaProperties,
+			env.clean(keySelector),
+			FlinkKafkaProducer.Semantic.EXACTLY_ONCE,
+			FlinkKafkaProducer.DEFAULT_KAFKA_PRODUCERS_POOL_SIZE);
+
+		// make sure the sink parallelism is set to producerParallelism
+		Preconditions.checkArgument(
+			kafkaProperties.getProperty(PRODUCER_PARALLELISM) != null,
+			"Missing producer parallelism for Kafka Shuffle");
+		int producerParallelism = PropertiesUtil.getInt(kafkaProperties, PRODUCER_PARALLELISM, Integer.MIN_VALUE);
+
+		addKafkaShuffle(inputStream, kafkaProducer, producerParallelism);
+	}
+
+	/**
+	 * Reads data from a Kafka Shuffle.
+	 *
+	 * <p>Consumers should read partitions equal to the key group indices they are assigned.
+	 * The number of partitions is equivalent to the key group sizes. Each consumer task reads from
+	 * one or multiple partitions. Any two consumer tasks can not read from the same partition.
+	 * Hence, the maximum parallelism of the receiving operator is the number of partitions.
+	 * This version only supports numberOfPartitions = consumerParallelism
```

Comment:
This is a good comment. Something like that should exist also for the other public API methods.